### PR TITLE
[DRAFT] Protocol redesign 

### PR DESCRIPTION
**Pull request by vmcru**
This pull request contains the integration of the new dataio files and paradigms. It looks to replace the bash files that run the experiment and the hyper-parameter optimization with new python files. The main changes pertain to the train.py file that loads the data using dataio files, new run_experiments.py file, and later a new run_hparam_optimization.py file that integrates with WANDB.

### CODE REVIEW

File: benchmarks/MOABB/dataio/preprocessing.py [hunk 1]
```diff
@@ -24,39 +24,82 @@ def to_tensor(epoch):
...
-        logging.warning(
-            "filter_length (%i) is longer than the signal (%i), "
-            "distortion is likely. Reduce filter length or filter a longer signal.",
-            filter_length,
-            len_x,
+def bandpass_resample():
+    @takes("epoch", "info", "target_sfreq", "fmin", "fmax")
+    @provides("epoch", "sfreq", "target_sfreq", "fmin", "fmax")
+    def _bandpass_resample(epoch, info, target_sfreq, fmin, fmax):
+        """Bandpass filter and resample an epoch."""
```
Comment (Drew-Wagner) [inline benchmarks/MOABB/dataio/preprocessing.py]: This is more what I had in mind as it makes it easy to configure from the hparams file. See how it is a function which creates a dynamic item with the desired parameters?

DONT commit the suggestion directly, as it is just to illustrate

```suggestion
def bandpass_resample(target_sfreq, fmin, fmax):
    @takes("epoch", "info")
    @provides("epoch")
    def _bandpass_resample(epoch, info):
        """Bandpass filter and resample an epoch."""
```

File: benchmarks/MOABB/hparams/MotorImagery/BNCI2014001/EEGNet.yaml [hunk 2]
```diff
@@ -39,6 +42,34 @@ C: 22
...
+# Create the subjects list
+#subjects: [!ref <target_subject>]
+
+# Create dataset using EpochedEEGDataset
+#dataset_class: !new:dataio.datasets.EpochedEEGDataset
+
+#json_path: !apply:os.path.join [!ref <cached_data_folder>, "index.json"]
+#save_path: !ref <data_folder>
+#dynamic_items:
+#    - !name:dataio.preprocessing.to_tensor
```
Comment (Drew-Wagner) [inline benchmarks/MOABB/hparams/MotorImagery/BNCI2014001/EEGNet.yaml]: This should include all preprocessing steps.

**Review by Drew-Wagner — COMMENTED** [review]
My initial thoughts after a 30 second skim... I did not check the experiment runner or train files yet.

File: benchmarks/MOABB/dataio/preprocessing.py (hunk 1 above)
Comment (Drew-Wagner) [reply benchmarks/MOABB/dataio/preprocessing.py]: This will allow us to do something like this in the yaml file:
```yaml
dynamic_items:
- !apply:bandpass_resample
  fmin: 0.5
  fmax: 22
  target_sfreq: 128
```

File: benchmarks/MOABB/hparams/MotorImagery/BNCI2014001/EEGNet.yaml [hunk 3]
```diff
@@ -165,3 +215,29 @@ model: !new:models.EEGNet.EEGNet
     dense_max_norm: !ref <dense_max_norm>
     dropout: !ref <dropout>
     dense_n_neurons: !ref <n_classes>
+
+# Search Space
```
Comment (pplantinga) [inline benchmarks/MOABB/hparams/MotorImagery/BNCI2014001/EEGNet.yaml:219]: I like the search space concept!

I wonder if there's a way to reference the search space in the rest of the yaml file -- like:

```
dropout: !ref <search_space.dropout>
```

Comment (pplantinga) [reply benchmarks/MOABB/hparams/MotorImagery/BNCI2014001/EEGNet.yaml:219]: This might make it slightly clearer with how it works rather than having to use overrides. You might still have to reload the file each time, depending on how the reference worked exactly.

File: benchmarks/MOABB/run_experiments.py [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+import string
+
+
+from train import load_hparams_and_prepare_data, run_experiment
+
+
+class ExperimentRunner:
+    """Manages multiple MOABB experiment runs."""
+
+    def __init__(self, args: list):
```
Comment (pplantinga) [inline benchmarks/MOABB/run_experiments.py:33]: maybe the `__init__` can take the relevant arguments directly and you can add a function like `from_commandline` to parse the command line and construct an object. e.g.

```python
def __init__(self, hparams, data_folder, output_folder, nsubj, nsess):
  """do input validation or whatever"""

@classmethod
def from_commandline_args(cls, args):
  """Load arguments from command line"""
  kwargs = self.parse_args(args)
  return cls(**kwargs)

def parse_args(self, args):
  parser = argparse.ArgumentParser()
  parser.add_argument(...)
  ...
  args = parser.parse_args(args)
  return args.__dict__
```

File: benchmarks/MOABB/train.py [hunk 5]
```diff
@@ -1,42 +1,181 @@
...
 
-    This class inherits from the Brain class in SpeechBrain.
-    The Brain class is the main class that handles training, validation,
-    testing, and checkpointing.
 
-    """
+class MOABBBrain(sb.Brain):
+    """Modified Brain class for MOABB experiments with new data format."""
 
     def init_model(self, model):
```
Comment (pplantinga) [inline benchmarks/MOABB/train.py:177]: Maybe model initialization should be done in the model files themselves?

File: benchmarks/MOABB/train.py [hunk 6]
```diff
@@ -253,174 +387,109 @@ def check_if_best(
 
 
 def run_experiment(hparams, run_opts, datasets):
-    """This function performs a single training (e.g., single cross-validation fold)"""
-    idx_examples = np.arange(datasets["train"].dataset.tensors[0].shape[0])
+    """Run a single experiment with the new data format."""
```
Comment (pplantinga) [inline benchmarks/MOABB/train.py:390]: "New" data format may not mean anything to someone reading this file. Perhaps you can either explicitly write out the data format, or just state more clearly what the function does.

File: benchmarks/MOABB/utils/search.py [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+def load_config_with_search_space(yaml_path):
+    """Load YAML config and return (base_config, search_space_dict)"""
+    with open(yaml_path) as f:
+        config = load_hyperpyyaml(f)  # <- THIS IS THE CORRECT WAY
+    base_config = {k: v for k, v in config.items() if k != "search_space"}
+    search_space = config.get("search_space", {})
+    return base_config, search_space
+
+
+def load_search_space_only(yaml_path):
```
Comment (pplantinga) [inline benchmarks/MOABB/utils/search.py:17]: I wonder if there's a better way to do this somehow. If the search space can't be integrated with the other parameters as suggested above, maybe this is just better as a separate file?

**Review by pplantinga — COMMENTED** [review]
Here are some initial comments from a cursory review. Things look generally good, but I wonder if the definition of the hyperparameter sweep configuration might be worth some further thought.

As noted in the meeting this could also support weights and biases.

File: benchmarks/MOABB/train.py [hunk 8]
```diff
@@ -1,42 +1,181 @@
...
+    overrides.update(
+        T=example_batch["epoch"].shape[1],  # Time dimension
+        C=example_batch["epoch"].shape[0],  # Channel dimension
+        n_train_examples=len(dataset),
+    )
+
+    # Reload hparams with shape information
+    with open(hparams_file) as fin:
+        hparams = load_hyperpyyaml(fin, overrides)
+
```
Comment (pplantinga) [inline benchmarks/MOABB/train.py:122]: I have found that the conversion of the dataset to a DynamicItemDataset results in a substantial slowdown due to the bandpass filter being applied to each sample on the CPU. One simple remedy  for small datasets could be to do the following:

```suggestion
    # Convert the dataset to a cached dataset
    dataset = InMemoryDataset(dataset)
```

Of course this requires an additional import at the beginning:
```python
from dataio.datasets import InMemoryDataset
```

File: benchmarks/MOABB/train.py [hunk 9]
```diff
@@ -253,174 +387,109 @@ def check_if_best(
...
     run_experiment(hparams, run_opts, datasets)
+    # except DryRunComplete:
+    #    print("Dry run successful")
+    #    sys.exit(0)
+    # except Exception as e:
+    #    print(f"Error during execution: {str(e)}")
+    #    if overrides.get("dry_run", False):
+    #        print("Dry run failed")
+    #        sys.exit(1)
+    #    raise
```
Comment (pplantinga) [inline benchmarks/MOABB/train.py:495]: Commented out blocks should be removed before merging.

File: benchmarks/MOABB/train.py [hunk 10]
```diff
@@ -62,58 +206,46 @@ def compute_forward(self, batch, stage):
...
-            + (1,)
-        )
-        model_summary = summary(
-            self.hparams.model, input_size=in_shape, device=self.device
-        )
-        with open(
-            os.path.join(self.hparams.exp_dir, "model.txt"), "w"
-        ) as text_file:
-            text_file.write(str(model_summary))
-
```
Comment (pplantinga) [inline benchmarks/MOABB/train.py:108]: I don't think this ought to be removed, it still seems useful. The `in_shape` will have to be changed though, the order of the two inner dimensions is reversed.

File: benchmarks/MOABB/train.py [hunk 11]
```diff
@@ -253,174 +387,109 @@ def check_if_best(
...
+    #            brain.compute_forward(batch, sb.Stage.TRAIN)
+    #        logger.info("✓ Dry run successful - model forward pass works")
+    #        raise DryRunComplete("Model validation successful")
+    #    except DryRunComplete:
+    #        raise
+    #    except Exception as e:
+    #        logger.error(f"✗ Dry run failed: {str(e)}")
+    #        raise
+
+    # Training
```
Comment (pplantinga) [inline benchmarks/MOABB/train.py:459]: Commented out code should be removed before merge

File: benchmarks/MOABB/train.py [hunk 12]
```diff
@@ -62,58 +206,46 @@ def compute_forward(self, batch, stage):
...
-        "Gets called when a stage (either training, validation, test) starts."
+        """Gets called when a stage (either training, validation, test) starts."""
         if stage != sb.Stage.TRAIN:
             self.preds = []
             self.targets = []
 
     def on_stage_end(self, stage, stage_loss, epoch=None):
         """Gets called at the end of a epoch."""
+        # Rest of the method remains the same as it handles metrics and checkpointing
+        # which don't need to change for the new data format
```
Comment (pplantinga) [inline benchmarks/MOABB/train.py:248]: "remains the same" as what? Whoever is reading this likely won't care what it looked like before.

File: benchmarks/MOABB/train.py [hunk 13]
```diff
@@ -1,42 +1,181 @@
...
-    > python train.py hparams/MotorImagery/BNCI2014001/EEGNet.yaml --data_folder=eeg_data --cached_data_folder=eeg_pickled_data --target_subject_idx=0 --target_session_idx=0 --data_iterator_name=leave-one-session-out
-
-Author
-------
-Davide Borra, 2022
-Mirco Ravanelli, 2023
+Authors
+-------
+Victor Cruz, 2025
+(Based on original work by Davide Borra and Mirco Ravanelli)
```
Comment (pplantinga) [inline benchmarks/MOABB/train.py:7]: Our typical practice in SpeechBrain so far has been to leave the original authors and just add new authors as they  edit the file.

**Review by pplantinga — COMMENTED** [review]
I have run the default recipe and there's a few small discrepencies with the original default recipe that seem to result in a bit worse accuracy. The ones I identified were the learning rate and the number of parameters. Let's make sure this achieves the same accuracy and runtime before merge.
